File: PP.Lab3C/PP.Lab3/Consumer.cs
using PoPLab3;

namespace PoPLab3
{
    public class Consumer
    {
        private readonly WorkInfo _info;
        private readonly BoundedStorage _storage;

        public Consumer(WorkInfo info, BoundedStorage storage)
        {
            _info = info;
            _storage = storage;
        }

        public void Run()
        {
            for (int i = 0; i < _info.Quota; i++)
            {
                Thread.Sleep(500);
                _storage.Take(_info.Id);
            }
        }
    }
}

File: PP.Lab3J/Consumer.java
import java.util.concurrent.CountDownLatch;

public class Consumer implements Runnable {
    private final WorkerInfo info;
    private final BoundedStorage storage;
    private final CountDownLatch latch;

    public Consumer(WorkerInfo info, BoundedStorage storage, CountDownLatch latch) {
        this.info = info;
        this.storage = storage;
        this.latch = latch;
    }

    @Override
    public void run() {
        try {
            for (int i = 1; i <= info.count(); i++) {
                storage.take(info.id()); 
                Thread.sleep(800); 
            }
            System.out.println("<-- Consumer " + info.id() + " has finished their quota.");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } finally {
            latch.countDown(); 
        }
    }
}